### DISCUSSION TITLE
EBPF program types should be restricted to their declared IRQL

### DISCUSSION THREAD

**Discussion by Alan-Jowett** [microsoft/ebpf-for-windows#2038] [Ideas]
eBPF program types should only be invoked at either passive or dispatch, but not both.

Some helper functions are only usable at either passive or dispatch. The list of helper functions usable for a specific program type should be restricted to only the helper functions whose IRQL requirements match the program types IRQL.

**Comment by saxena-anurag** [comment]
Per triage meeting discussion:

We should have a `max IRQL` level for each program type. For example, SOCK_ADDR program type will have max IRQL as DISPATCH, and then SOCK_ADDR hooks can invoke programs at either passive or dispatch.

What helper functions can be invoked by the program type will depend on the `max IRQL` level.
